packages/playwright-core/src/server/fetch.ts

```diff
@@ -422,14 +422,35 @@ export abstract class APIRequestContext extends SdkObject {
 
         const chunks: Buffer[] = [];
         const notifyBodyFinished = () => {
+          function relativeTime(time: number | undefined): number {
+            if (!time)
+              return -1;
+            return time - startAt;
+          }
+          const endAt = monotonicTime();
+          // spec: https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming
+          const timing: channels.ResourceTiming = {
+            startTime: startAt,
+            domainLookupStart: dnsLookupAt ? 0 : -1,
+            domainLookupEnd: dnsLookupAt ? dnsLookupAt! - startAt : -1,
+            connectStart: dnsLookupAt ? dnsLookupAt! - startAt  : 0,
+            secureConnectionStart: dnsLookupAt ? dnsLookupAt! - startAt : 0,
+            connectEnd: relativeTime(tlsHandshakeAt ?? tcpConnectionAt),
+            requestStart: relativeTime(tlsHandshakeAt ?? tcpConnectionAt),
+            responseStart: responseAt - startAt,
+          };
+          const responseEndTiming = endAt - startAt;
+
           const body = Buffer.concat(chunks);
           notifyRequestFinished(body);
           fulfill({
             url: response.url || url.toString(),
             status: response.statusCode || 0,
             statusText: response.statusMessage || '',
             headers: toHeadersArray(response.rawHeaders),
-            body
+            body,
+            timing,
+            responseEndTiming,
```

Comment:
Why do we have two sets of timings now?

Reply:
For browser requests, most of the timings are transmitted in the `Response` type, and then `responseEndTiming` arrives later in the `requestFinished` and `requestFailed` events. I opted to make this similar, so we have a big set of timings in `timing` and the final timing in `responseEndTiming`.

I thought about amending the `ResourceTiming` type instead, but then `requestFinished` would suddenly have the response end time both in `responseEndTiming` and in `response.timing.responseEnd` - that felt confusing.